packages/itwin/tree-widget/src/components/trees/models-tree/HierarchyBasedVisibilityHandler.ts

```diff
@@ -0,0 +1,848 @@
+          return EMPTY;
+        }
+
+        return concat(
+          props.on && !viewport.view.viewsModel(modelId) ? viewport.addViewedModels(modelId) : EMPTY,
+          defer(() => {
+            const isDisplayedInSelector = viewport.view.viewsCategory(categoryId);
+            const override =
+              on === isDisplayedInSelector
+                ? PerModelCategoryVisibility.Override.None
+                : on
+                  ? PerModelCategoryVisibility.Override.Show
+                  : PerModelCategoryVisibility.Override.Hide;
+            viewport.perModelCategoryVisibility.setOverride(modelId, categoryId, override);
+            if (override === PerModelCategoryVisibility.Override.None && on) {
+              // we took off the override which means the category is displayed in selector, but
+              // doesn't mean all its subcategories are displayed - this call ensures that
+              viewport.changeCategoryDisplay(categoryId, true, true);
+            }
+            return EMPTY;
+          }),
+        );
+      }),
+      defer(() => this.clearAlwaysAndNeverDrawnChildren(props)),
+    );
+
+    const ovr = this._props.overrides?.changeCategoryState;
+    return ovr ? from(ovr(this.createVoidOverrideProps(props, result))) : result;
+  }
+
+  /**
+   * Updates visibility of all grouping node's elements.
+   * @see `changeElementState`
+   */
+  private changeElementGroupingNodeState(key: ECClassGroupingNodeKey, on: boolean): Observable<void> {
+    const result = defer(() => {
+      return this._queryHandler.queryGroupingNodeChildren(key).pipe(
+        mergeMap(({ modelId, categoryId, elementIds }) => {
+          const viewport = this._props.viewport;
+          return concat(
+            on && !viewport.view.viewsModel(modelId) ? from(viewport.addViewedModels(modelId)) : EMPTY,
+            defer(() => {
+              const categoryVisibility = this.getDefaultCategoryVisibilityStatus(categoryId, modelId);
+              const isDisplayedByDefault = categoryVisibility.state === "visible";
+              return from(elementIds).pipe(this.changeElementStateNoChildrenOperator({ on, isDisplayedByDefault }));
+            }),
+          );
+        }),
+      );
+    });
+
+    const ovr = this._props.overrides?.changeElementGroupingNodeState;
+    return ovr ? from(ovr(this.createVoidOverrideProps({ key, on }, result))) : result;
+  }
+
+  /**
+   * Updates visibility of an element and all its child elements by adding them to the always/never drawn list.
+   * @note If element is to be enabled and model is hidden, it will be enabled.
+   */
+  private changeElementState(props: ChangeElementStateProps): Observable<void> {
+    const result = defer(() => {
+      const { elementId, on, hasChildren, modelId, categoryId } = props;
+      const viewport = this._props.viewport;
+      return concat(
+        props.on && !viewport.view.viewsModel(modelId) ? from(viewport.addViewedModels(modelId)) : EMPTY,
```

Comment:
This might enable display of all model's elements, although we only want to turn on one element (and its children). If the model was turned off, all its elements were 100% hidden. Based on that, we could:
1. set per-model category overrides to "off" for every model category,
2. ensure that always drawn list doesn't contain any model elements.

Reply:
You are right, I kind of missed this case. Added suggested changes